File: Algorithms/813/largest-sum-of-averages.cpp
class Solution {
public:
    double largestSumOfAverages(vector<int>& A, int K) {
        double dp[105][105];
        memset(dp, 0, sizeof dp);
        
        int sum = 0, cnt = 0, len = (int)A.size() - K + 1;
        for (int i = 0; i < len; ++i) {
            sum += A[i];
            dp[0][i] = (double)sum/++cnt;
        }
        
        for (int ik = 1; ik < K; ++ik) {
            len = (int)A.size() - K + ik + 1;
            for (int st = ik; st < len; ++st) {
                sum = cnt = 0;
                for (int ed = st; ed < len; ++ed) {
                    sum += A[ed];
                    auto avg = (double)sum/++cnt;
                    dp[ik][ed] = max(dp[ik][ed], dp[ik-1][st-1] + avg);
                }
            }
        }
        
        return dp[K-1][A.size()-1];
    }
};

File: Algorithms/[1, 1k)/813/largest-sum-of-averages.swift
class Solution {
    func largestSumOfAverages(_ A: [Int], _ K: Int) -> Double {
        var dp = Array.init(repeating: Array.init(repeating: 0.0, count: A.count), count: K)
        
        var sum = 0
        var cnt = 0
        var len = A.count - K + 0 + 1
        for i in 0..<len {
            sum += A[i]
            cnt += 1
            dp[0][i] = Double(sum)/Double(cnt)
        }
        
        for ik in 1..<K {
            len = A.count - K + ik + 1
            for st in ik..<len {// [st, ed]
                sum = 0
                cnt = 0
                for ed in st..<len {
                    sum += A[ed]
                    cnt += 1
                    let avg = Double(sum)/Double(cnt)
                    dp[ik][ed] = max(dp[ik][ed], dp[ik-1][st-1] + avg)
                }
            }
        }
        
        return dp[K - 1][A.count - 1]
    }
}